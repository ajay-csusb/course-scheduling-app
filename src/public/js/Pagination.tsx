import React, { Component } from 'react';
import * as ClassSearchUtils from './ClassSearchUtils';

export interface IPaginationProps {
  numberPages: number;
  onChangeOfPageNumber: (event: any) => void;
  currentPage: number;
}

export default class Pagination extends Component<IPaginationProps, {}> {
  constructor(props: IPaginationProps) {
    super(props);
    this.handleOnChangeOfPageNumber = this.handleOnChangeOfPageNumber.bind(this);
  }

  public render(): React.ReactNode {
    let prevPage = <></>;
    let nextPage = <></>;
    let firstPage = <></>;
    let lastPage = <></>;
    let pages: JSX.Element[] = [];
    const firstPageNumber = 1;
    const lastPageNumber = this.props.numberPages;
    const prevPageNumber = this.props.currentPage !== 1 ? this.props.currentPage - 1 : null;
    const nextPageNumber = this.props.currentPage !== this.props.numberPages ? this.props.currentPage + 1 : null;
    if (this.props.numberPages > 1) {
      prevPage = prevPageNumber ? this.getPagerMarkup(prevPageNumber, 'Previous', 0) : <></>;
      nextPage = nextPageNumber ? this.getPagerMarkup(nextPageNumber, 'Next', this.props.currentPage + 1) : <></>;
      firstPage = this.getPagerMarkup(firstPageNumber, 'First', firstPageNumber - 2);
      lastPage = this.getPagerMarkup(lastPageNumber, 'Last', lastPageNumber + 2);
    }
    for (let index = 0; index < this.props.numberPages; index++) {
      const pageNumber = index + 1;
      pages.push(this.getPagerMarkup(pageNumber, pageNumber.toString(), pageNumber));
    }
    pages = ClassSearchUtils.updatePages(pages, this.props.currentPage);
    return (
      <div>
        <ul className="pagination">
          {firstPage}
          {prevPage}
          {pages}
          {nextPage}
          {lastPage}
        </ul>
      </div>
    );
  }

  private handleOnChangeOfPageNumber(event: any): void {
    const currentPageNumber = parseInt(event.target.getAttribute('data-page'), 10);
    this.props.onChangeOfPageNumber(currentPageNumber);
  }

  private getPagerMarkup(pageNumber: number, label: string, key: number): JSX.Element {
    const cssClass = this.props.currentPage === pageNumber ? 'current page-number' : 'page-number';
    return (
      <li className={`page-${pageNumber}`} key={key}>
        <a
          title={`Go to page ${pageNumber}`}
          onClick={this.handleOnChangeOfPageNumber}
          data-page={`${pageNumber}`}
          className={cssClass}
          href='#class-search-results-component'
        >
          {label}
        </a>
      </li>
    );
  }
}
