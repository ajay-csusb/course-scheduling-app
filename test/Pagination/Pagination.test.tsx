import { mount } from 'enzyme';
import * as React from 'react';
import Pagination from '../../src/public/js/Pagination';
import { TestUtils } from '../TestUtils';
// tslint:disable:max-line-length
describe('Pagination component', () => {
  beforeAll(() => {
    TestUtils.ajax();
  });

  test('pagination component snapshot', () => {
    const paginationComponentWrapper = mount(
      <Pagination numberPages={10} onChangeOfPageNumber={() => {}} currentPage={1} />
    );
    expect(paginationComponentWrapper).toMatchSnapshot();
  });

  describe('when a pager is displayed', () => {
    const paginationComponentWrapper = mount(
      <Pagination numberPages={10} onChangeOfPageNumber={() => {}} currentPage={2} />
    );
    it('should display the correct number of pages', () => {
      expect(paginationComponentWrapper.find('.pagination li a')).toHaveLength(14);
    });

    it('should display the first and last page link', () => {
      expect(paginationComponentWrapper.find('.pagination').html()).toMatch(/First/);
      expect(paginationComponentWrapper.find('.pagination').html()).toMatch(/Last/);
    });

    it('should display the previous and next link', () => {
      expect(paginationComponentWrapper.find('.pagination').html()).toMatch(/Previous/);
      expect(paginationComponentWrapper.find('.pagination').html()).toMatch(/Next/);
    });

    it('should display the correct CSS class', () => {
      expect(paginationComponentWrapper.find('.pagination li a').at(3).html()).toMatch(/current page-number/);
    });

    describe('when we are on the first page', () => {
      it('should not display the previous link', () => {
        const paginationComponentWrapper = mount(
          <Pagination numberPages={10} onChangeOfPageNumber={() => {}} currentPage={1} />
        );
        expect(paginationComponentWrapper.find('.pagination').html()).not.toMatch(/Previous/);
      });
    });

    describe('when we are on the last page', () => {
      it('should not display the next link', () => {
        const paginationComponentWrapper = mount(
          <Pagination numberPages={10} onChangeOfPageNumber={() => {}} currentPage={10} />
        );
        expect(paginationComponentWrapper.find('.pagination').html()).not.toMatch(/Next/);
      });
    });
  });
});
