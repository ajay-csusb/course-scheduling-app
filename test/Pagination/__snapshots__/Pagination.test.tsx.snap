// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component pagination component snapshot 1`] = `
<Pagination
  currentPage={1}
  numberPages={10}
  onChangeOfPageNumber={[Function]}
>
  <div>
    <ul
      className="pagination"
    >
      <li
        className="page-1"
        key="-1"
      >
        <a
          className="current page-number"
          data-page="1"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 1"
        >
          First
        </a>
      </li>
      <li
        className="page-1"
        key="1"
      >
        <a
          className="current page-number"
          data-page="1"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 1"
        >
          1
        </a>
      </li>
      <li
        className="page-2"
        key="2"
      >
        <a
          className="page-number"
          data-page="2"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 2"
        >
          2
        </a>
      </li>
      <li
        className="page-3"
        key="3"
      >
        <a
          className="page-number"
          data-page="3"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 3"
        >
          3
        </a>
      </li>
      <li
        className="page-4"
        key="4"
      >
        <a
          className="page-number"
          data-page="4"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 4"
        >
          4
        </a>
      </li>
      <li
        className="page-5"
        key="5"
      >
        <a
          className="page-number"
          data-page="5"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 5"
        >
          5
        </a>
      </li>
      <li
        className="page-6"
        key="6"
      >
        <a
          className="page-number"
          data-page="6"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 6"
        >
          6
        </a>
      </li>
      <li
        className="page-7"
        key="7"
      >
        <a
          className="page-number"
          data-page="7"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 7"
        >
          7
        </a>
      </li>
      <li
        className="page-8"
        key="8"
      >
        <a
          className="page-number"
          data-page="8"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 8"
        >
          8
        </a>
      </li>
      <li
        className="page-9"
        key="9"
      >
        <a
          className="page-number"
          data-page="9"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 9"
        >
          9
        </a>
      </li>
      <li
        className="page-10"
        key="10"
      >
        <a
          className="page-number"
          data-page="10"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 10"
        >
          10
        </a>
      </li>
      <li
        className="page-2"
        key="2"
      >
        <a
          className="page-number"
          data-page="2"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 2"
        >
          Next
        </a>
      </li>
      <li
        className="page-10"
        key="12"
      >
        <a
          className="page-number"
          data-page="10"
          href="#class-search-results-component"
          onClick={[Function]}
          title="Go to page 10"
        >
          Last
        </a>
      </li>
    </ul>
  </div>
</Pagination>
`;
